File: csharp/src/Training.Workshop/M8/S07Adr/Step1/BookingService.cs
using System.Globalization;
using Training.Workshop.M8.S07Adr.Step1.Notification;
using Training.Workshop.M8.S07Adr.Step1.Pricing;

namespace Training.Workshop.M8.S07Adr.Step1;

/// <summary>Krok 1: warstwa aplikacji przejmuje wysłanie powiadomienia na podstawie Quote.</summary>
public sealed class BookingService
{
    private readonly GroupMailer _mailer = new();
    private readonly TicketPricing _pricing = new();

    public string Book(string organizer, int tickets, string format)
    {
        double unitPrice = format switch
        {
            "IMAX" => 40.00,
            "3D" => 32.00,
            _ => 25.00,
        };
        Quote quote = _pricing.Total(tickets, unitPrice);
        if (quote.GroupDiscount)
        {
            _mailer.GroupDiscountGranted(organizer, tickets);
        }
        return "DO ZAPLATY " + quote.Total.ToString("F2", CultureInfo.InvariantCulture);
    }

    public IReadOnlyList<string> SentMails()
    {
        return _mailer.Sent();
    }
}

File: typescript/src/workshop/m8/s07_adr/step1/BookingService.ts
import { GroupMailer } from './notification/GroupMailer.js';
import { TicketPricing } from './pricing/TicketPricing.js';

/** Krok 1: warstwa aplikacji przejmuje wysłanie powiadomienia na podstawie Quote. */
export class BookingService {
  private readonly mailer = new GroupMailer();
  private readonly pricing = new TicketPricing();

  book(organizer: string, tickets: number, format: string): string {
    const unitPrice = BookingService.unitPrice(format);
    const quote = this.pricing.total(tickets, unitPrice);
    if (quote.groupDiscount) {
      this.mailer.groupDiscountGranted(organizer, tickets);
    }
    return 'DO ZAPLATY ' + quote.total.toFixed(2);
  }

  sentMails(): readonly string[] {
    return this.mailer.sent();
  }

  private static unitPrice(format: string): number {
    switch (format) {
      case 'IMAX': return 40.00;
      case '3D': return 32.00;
      default: return 25.00;
    }
  }
}
